### ISSUE TITLE
ClassCastException thrown on Scala 3 when transforming sealed traits to Protobuf GeneratedEnum

### ISSUE THREAD

**Issue by saeltz** [scalalandio/chimney#479]
**Checklist**

- [x] I read the documentation at https://chimney.readthedocs.io/ and checked that the functionality exists
- [x] I verified that the behavior for my use case doesn't match the documentation
- [x] I checked the https://github.com/scalalandio/chimney/issues and haven't found the issue reported
  - [x] I checked the https://github.com/scala/bug/issues and haven't found the issue reported there - relevant for Scala 2 compiler crashes, since it might be a bug in the compiler
  - [x] I checked the https://github.com/lampepfl/dotty/issues and haven't found the issue reported there - relevant for Scala 3 compiler crashes, since it might be a bug in the compiler 
- [x] I confirmed that the bug is not related to functionality that was deprecated: lifted transformers (`TransformerF`s) or `unsafeOption` flags
- [x] I believe that this is actual bug, rather than [a question how to use existing features](https://github.com/scalalandio/chimney/discussions)

**Describe the bug**

A `ClassCastException` is thrown when trying to transform a `sealed trait` into a Protobuf `GeneratedEnum`.

**Reproduction**

Run [the test at this commit of this repository](https://github.com/moia-oss/teleproto/blob/f72e40a521f7f6a02b5728db340dc9d9f0913243/src/test/scala/io/moia/protos/teleproto/ProtocolBuffersRoundTripTest.scala).

This doesn't necessarily happen at every run but on most of the runs. Repeatedly `clean` before running the test may yield a different result.

**Expected behavior**

No exception.

**Actual behavior**

```
java.lang.ClassCastException: class io.moia.protos.teleproto.ProtocolBuffersRoundTripTest$Color$Yellow$ cannot be cast to class io.moia.protos.teleproto.ProtocolBuffersRoundTripTest$Color$pink$ (io.moia.protos.teleproto.ProtocolBuffersRoundTripTest$Color$Yellow$ and io.moia.protos.teleproto.ProtocolBuffersRoundTripTest$Color$pink$ are in unnamed module of loader sbt.internal.LayeredClassLoader @7dfaa86d)
[info]   at io.moia.protos.teleproto.ProtocolBuffersRoundTripTest$$anon$2.transform(ProtocolBuffersRoundTripTest.scala:20)
[info]   at io.moia.protos.teleproto.ProtocolBuffersRoundTripTest$$anon$2.transform(ProtocolBuffersRoundTripTest.scala:37)
[info]   at io.moia.protos.teleproto.ProtocolBuffersRoundTripTest$$anon$1.transform$$anonfun$1(ProtocolBuffersRoundTripTest.scala:46)
...
```

What's especially weird is that the first line of the stack trace is commented out.

**Which Chimney version do you use**

1.0.0-M1

**Which platform do you use**

- [x] JVM
- [ ] Scala.js
- [ ] Scala Native

**If you checked JVM**

OpenJDK 21.0.2

**Additional context**

We reproduced that on two MacBooks Pro M1.

If you go to [the next commit of the branch](https://github.com/moia-oss/teleproto/blob/f8b92dc00df8c578b91421f9617974a11df72d41/src/test/scala/io/moia/protos/teleproto/ProtocolBuffersRoundTripTest.scala) I fixed the exception by converting the `sealed trait` into a Scala 3 `enum`. I still thought it'd be helpful for you to take a look.

**Comment by MateuszKubuszok** [comment]
Thank you for the report!

As I wrote in #480, I don't have much time on my hand so I am not sure when I'll be able to look at the reproduction.

If you wanted to investivate further in the meantime

 * it could be interesting to [take a look at the generated code](https://chimney.readthedocs.io/en/stable/troubleshooting/#debugging-macros), to see if handwritten version would have the same issue

 * it's very curious , if `clean` and `compile` can change the outcome, as it suggest it might be something about Zink and/or classloader. I started observing many issues ever since I updated Scala 3 to 3.3.3 - on 3.3.1 code was compiling fine, but after update I get a lot of broken classpath exceptions, and I have to `compile` sometimes 3 times or more before it works. I was thinking it might be an artifact of some weird changes to build.sbt I attempted a few days ago, but maybe it's related to the Scala 3 update. I see you have 3.3.3 as well. What would happen if you downgrade to 3.3.1?

**Comment by saeltz** [comment]
The erroneous derived transformer looks like:

```
[info]    || {
[info]    ||   final class $anon() extends io.scalaland.chimney.Transformer[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color] {
[info]    ||     def transform(src: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color): io.moia.food.food.Meal.Color = {
[info]    ||       val color: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color = src
[info]    ||       (color: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color @scala.unchecked) match {
[info]    ||         case pink @ io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.pink =>
[info]    ||           val pink: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.pink.type = pink
[info]    ||           {
[info]    ||             (pink: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.pink @scala.unchecked) match {
[info]    ||               case _ =>
[info]    ||                 ()
[info]    ||             }
[info]    ||             ()
[info]    ||           }
[info]    ||           vector.apply(1).asInstanceOf[scala.Function1[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color]].apply(pink.asInstanceOf[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color])
[info]    ||         case red @ io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Red =>
[info]    ||           val red: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Red.type = red
[info]    ||           {
[info]    ||             (red: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Red @scala.unchecked) match {
[info]    ||               case _ =>
[info]    ||                 ()
[info]    ||             }
[info]    ||             ()
[info]    ||           }
[info]    ||           vector.apply(3).asInstanceOf[scala.Function1[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color]].apply(red.asInstanceOf[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color])
[info]    ||         case yellow @ io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Yellow =>
[info]    ||           val yellow: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Yellow.type = yellow
[info]    ||           {
[info]    ||             (yellow: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Yellow @scala.unchecked) match {
[info]    ||               case _ =>
[info]    ||                 ()
[info]    ||             }
[info]    ||             ()
[info]    ||           }
[info]    ||           vector.apply(4).asInstanceOf[scala.Function1[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color]].apply(yellow.asInstanceOf[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color])
[info]    ||         case blue @ io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Blue =>
[info]    ||           val blue: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Blue.type = blue
[info]    ||           {
[info]    ||             (blue: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Blue @scala.unchecked) match {
[info]    ||               case _ =>
[info]    ||                 ()
[info]    ||             }
[info]    ||             ()
[info]    ||           }
[info]    ||           vector.apply(0).asInstanceOf[scala.Function1[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color]].apply(blue.asInstanceOf[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color])
[info]    ||         case orange @ io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.orange =>
[info]    ||           val orange: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.orange.type = orange
[info]    ||           {
[info]    ||             (orange: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.orange @scala.unchecked) match {
[info]    ||               case _ =>
[info]    ||                 ()
[info]    ||             }
[info]    ||             ()
[info]    ||           }
[info]    ||           vector.apply(2).asInstanceOf[scala.Function1[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color]].apply(orange.asInstanceOf[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color])
[info]    ||       }
[info]    ||     }
[info]    ||   }
[info]    || 
[info]    ||   (new $anon(): io.scalaland.chimney.Transformer[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color])
[info]    || }
```

The enum one:

```
[info]    || {
[info]    ||   final class $anon() extends io.scalaland.chimney.Transformer[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color] {
[info]    ||     def transform(src: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color): io.moia.food.food.Meal.Color = {
[info]    ||       val color: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color = src
[info]    ||       (color: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color @scala.unchecked) match {
[info]    ||         case blue: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Blue.type =>
[info]    ||           val blue: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Blue.type = blue
[info]    ||           {
[info]    ||             (blue: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Blue @scala.unchecked) match {
[info]    ||               case _ =>
[info]    ||                 ()
[info]    ||             }
[info]    ||             ()
[info]    ||           }
[info]    ||           vector.apply(0).asInstanceOf[scala.Function1[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color]].apply(blue.asInstanceOf[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color])
[info]    ||         case yellow: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Yellow.type =>
[info]    ||           val yellow: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Yellow.type = yellow
[info]    ||           {
[info]    ||             (yellow: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Yellow @scala.unchecked) match {
[info]    ||               case _ =>
[info]    ||                 ()
[info]    ||             }
[info]    ||             ()
[info]    ||           }
[info]    ||           vector.apply(4).asInstanceOf[scala.Function1[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color]].apply(yellow.asInstanceOf[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color])
[info]    ||         case red: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Red.type =>
[info]    ||           val red: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Red.type = red
[info]    ||           {
[info]    ||             (red: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.Red @scala.unchecked) match {
[info]    ||               case _ =>
[info]    ||                 ()
[info]    ||             }
[info]    ||             ()
[info]    ||           }
[info]    ||           vector.apply(3).asInstanceOf[scala.Function1[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color]].apply(red.asInstanceOf[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color])
[info]    ||         case pink: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.pink.type =>
[info]    ||           val pink: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.pink.type = pink
[info]    ||           {
[info]    ||             (pink: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.pink @scala.unchecked) match {
[info]    ||               case _ =>
[info]    ||                 ()
[info]    ||             }
[info]    ||             ()
[info]    ||           }
[info]    ||           vector.apply(1).asInstanceOf[scala.Function1[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color]].apply(pink.asInstanceOf[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color])
[info]    ||         case orange: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.orange.type =>
[info]    ||           val orange: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.orange.type = orange
[info]    ||           {
[info]    ||             (orange: io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color.orange @scala.unchecked) match {
[info]    ||               case _ =>
[info]    ||                 ()
[info]    ||             }
[info]    ||             ()
[info]    ||           }
[info]    ||           vector.apply(2).asInstanceOf[scala.Function1[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color]].apply(orange.asInstanceOf[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color])
[info]    ||       }
[info]    ||     }
[info]    ||   }
[info]    || 
[info]    ||   (new $anon(): io.scalaland.chimney.Transformer[io.moia.protos.teleproto.ProtocolBuffersRoundTripTest.Color, io.moia.food.food.Meal.Color])
[info]    || }
```

Ignoring all the boilerplate, the only difference is the colon `:` instead of an `@` in each `case` of the pattern match.

I don't really know the value of `vector` and can't find it. I assume it's a `Vector[io.moia.food.food.Meal.Color]` where the `To` type is gotten by index. Is there a way to get that? I could then try to use the derived code and debug further.

**Comment by saeltz** [comment]
Using Scala 3.3.1 the problem still exists. Also on Scala 3.4. (Although I found another issue with ScalaPB: https://github.com/scalapb/ScalaPB/issues/1662)

**Comment by saeltz** [comment]
I minimised the example even further in https://github.com/moia-oss/teleproto/blob/chimney2/src/test/scala/io/moia/protos/teleproto/ProtocolBuffersRoundTripTest.scala.

I found another difference: It's not only the `:` but also the `.type` in the same line.

The `ClassCastException` is thrown by the `@unchecked` annotation as documented here:
https://www.scala-lang.org/api/current/scala/unchecked.html. I don't really understand why it's not thrown in the `enum` variant.

**Comment by jchyb** [comment]
I looked at this yesterday and it seems that the issue is not with the pattern match case catching incorrect values but with incorrect `runtimeData` being used in the rhs there. E.g.:
```scala
//> using dep "io.scalaland::chimney:0.8.5"

import io.scalaland.chimney.dsl.*
import io.scalaland.chimney.Transformer

sealed trait Color
object Color {
  case object orange extends Color
  case object pink extends Color
}

sealed trait Target
object Target {
  case class Impl(value: String) extends Target
}

@main def main() = 
  val writer: Transformer[Color, Target] = Transformer
    .define[Color, Target]
    .withCoproductInstance[Color.orange.type](_ => Target.Impl("orange"))
    .withCoproductInstance[Color.pink.type](_ => Target.Impl("pink"))
    .enableMacrosLogging
    .buildTransformer

  writer.transform(Color.pink)
```
when running this with `-Xprint:inlining`, we get (I trimmed some of the types for readability):
<details>

```scala
val writer: io.scalaland.chimney.Transformer[Color, Target] =
          {
            val TransformerDefinition_this:  io.scalaland.chimney.dsl.TransformerDefinition[...]
             = io.scalaland.chimney.Transformer.define[Color, Target]
            val TransformerDefinition_this: 
                io.scalaland.chimney.dsl.TransformerDefinition[...]
             =
              io.scalaland.chimney.internal.runtime.WithRuntimeDataStore.update[ // idx: 0, orange
               ...
              ](TransformerDefinition_this,
                {
                  def $anonfun(_$1: Color.orange.type): Target =
                    Target.Impl.apply("orange")
                  closure($anonfun)
                }
              ).asInstanceOf[...]
            {
              val TransformerDefinition_this:
                  io.scalaland.chimney.dsl.TransformerDefinition[...]
               =
                io.scalaland.chimney.internal.runtime.WithRuntimeDataStore.
                  update[...](TransformerDefinition_this, // idx: 1, pink
                  {
                    def $anonfun(_$2: Color.pink.type): Target =
                      Target.Impl.apply("pink")
                    closure($anonfun)
                  }
                ).asInstanceOf[...].enableMacrosLogging
              {
                val vector$macro$1: Vector[Any] =
                  TransformerDefinition_this.runtimeData
                {
                  {
                    (vector$macro$1:Vector[Any] @unchecked match 
                      {
                        case _ => ()
                      }
                    )
                    ()
                  }
                  {
                    final class $anon() extends Object(), io.scalaland.chimney.
                      Transformer[Color, Target] {
                      def transform(src: Color): Target =
                        {
                          val color$macro$1: Color = src
                          color$macro$1:Color @unchecked match 
                            {
                              case orange$macro$2 @ Color.orange =>
                                val orange$macro$1: Color.orange.type =
                                  orange$macro$2
                                {
                                  orange$macro$1:Color.orange.type @unchecked
                                     match 
                                    {
                                      case _ => ()
                                    }
                                  ()
                                }
                                vector$macro$1.apply(1).asInstanceOf[ // idx 1: pink for type orange
                                  Color => Target].apply(
                                  orange$macro$1.asInstanceOf[Color])
                              case pink$macro$2 @ Color.pink =>
                                val pink$macro$1: Color.pink.type = pink$macro$2
                                {
                                  pink$macro$1:Color.pink.type @unchecked match 
                                    {
                                      case _ => ()
                                    }
                                  ()
                                }
                                vector$macro$1.apply(0).asInstanceOf[ // idx: 0, orange, for type pink
                                  Color => Target].apply(
                                  pink$macro$1.asInstanceOf[Color])
                            }
                        }
                    }
                    new $anon():io.scalaland.chimney.Transformer[Color, Target]
                  }
                }
              }:io.scalaland.chimney.Transformer[Color, Target]
            }
          }
```

</details>

This can also be replicated in tests, however changing the order of `withCoproductInstance` will "fix" the issue, so I imagine that is why it was not discovered there. I have not managed to fix this yet, unfortunately

**Comment by MateuszKubuszok** [comment]
@jchyb thank you for figuring it out!

If that's the case then either:

 * there is a bug in how indices are calculated in https://github.com/scalalandio/chimney/blob/master/chimney/src/main/scala/io/scalaland/chimney/internal/compiletime/derivation/transformer/Configurations.scala#L291 or
 * there is bug in how configs are matched against values in https://github.com/scalalandio/chimney/blob/master/chimney/src/main/scala/io/scalaland/chimney/internal/compiletime/derivation/transformer/rules/TransformSealedHierarchyToSealedHierarchyRuleModule.scala#L60

So to fix it we'd have to:

 - [ ] create reproduction in https://github.com/scalalandio/chimney/blob/master/chimney/src/test/scala/io/scalaland/chimney/IssuesSpec.scala
   - [ ] it might be a good idea to come up with several similar cases, e.g. with `withCoproductInstant`, `withCoproductInstantPartial`, maybe also something similar for case classes and fields
   - [ ] the Scastie code might be a good starting point
 - [ ] use `enableMacrosLogging` to figure out which part of the code maps indices incorrectly
 - [ ] fix it and verify that tests are passing

@saeltz would you be interested in trying to fix it?

**Comment by saeltz** [comment]
> would you be interested in trying to fix it?

I first want to get #489 over the finish line.

**Comment by MateuszKubuszok** [comment]
@jchyb I believe that it a bug in [pattern matching on Scala 3](https://github.com/scalalandio/chimney/blob/6f4390ad80c2a2faf521332cc81e8d775c89da9e/chimney-macro-commons/src/main/scala-3/io/scalaland/chimney/internal/compiletime/ExprPromisesPlatform.scala#L101) after all.

I compared:

 * type level representations of config
 * coproductOverrides

in

 * Scala 2.13 - https://scastie.scala-lang.org/IDPfJJBLRBmoNhmnrGseQg
 * Scala 3 - https://scastie.scala-lang.org/5oxrmiI8QkGMENM0qzo5qg

They are correct - runtime values are prepended so that what is in `Vector[Any]` corresponds with the position with type-level-list of overrides (rename is an exception but in everything else prepends both on type-level and on value-level).

What changes is the order of `case` clauses in a pattern matching - at first Chimney generates them from CoproductInstance overrides - in the order it found them in a config, so it's deterministic - and then fill the remaining subtypes by matching names.

You can see that failing example has `orange` as the first `case` and working one has `pink` as the first `case`. Pass `orange` and the outcome reverses.

(I had a much longer comment but I accidentally refreshed and it all got removed :| )

**Comment by MateuszKubuszok** [comment]
Since there is nothing we can do do fix this, and the only solution I can see is fixing it in the compiler, I'll close it now.

**Comment by MateuszKubuszok** [comment]
Reported as https://github.com/scala/scala3/issues/20350

**Comment by saeltz** [comment]
It seems to be an issue with the macro itself, see https://github.com/scala/scala3/issues/20350#issuecomment-2368399331. Would you consider re-opening and fixing?

**Comment by MateuszKubuszok** [comment]
I can reopen it, but I don't see any capacity for doing much OSS work anytime soon in my schedule.

**Comment by MateuszKubuszok** [comment]
Released in [1.5.0](https://github.com/scalalandio/chimney/releases/tag/1.5.0)
